File: src/year2024/day15/day15.rs
//! [Day 15: Warehouse Woes](https://adventofcode.com/2024/day/15)

use rustc_hash::FxHashSet;
use std::collections::VecDeque;

use aoc::Coord;

type Grid = aoc::Grid<char>;

fn score(grid: &Grid) -> i32 {
    grid.iter_cells()
        .filter(|(_, &c)| c == 'O' || c == '[')
        .map(|(xy, _)| 100 * xy.y + xy.x)
        .sum()
}

fn init_first_warehouse(input: &str) -> (Grid, Coord) {
    let mut grid = Grid::parse(input);
    let mut start = Coord { x: 0, y: 0 };

    for (pos, &c) in &grid {
        if c == '@' {
            start = pos;
            break;
        }
    }
    grid[start] = '.';

    (grid, start)
}

fn init_second_warehouse(input: &str) -> (Grid, Coord) {
    let simple = Grid::parse(input);
    let mut grid = Grid::with_size(simple.width() * 2, simple.height(), ' ', '#');
    let mut start = Coord::new(0, 0);

    for (Coord { x, y }, &c) in simple.iter_cells() {
        let pos1 = Coord { x: x * 2, y };
        let pos2 = Coord { x: x * 2 + 1, y };
        match c {
            '@' => {
                start = pos1;
                grid[pos1] = '.';
                grid[pos2] = '.';
            }
            'O' => {
                grid[pos1] = '[';
                grid[pos2] = ']';
            }
            _ => {
                grid[pos1] = c;
                grid[pos2] = c;
            }
        };
    }

    (grid, start)
}

fn move_boxes(grid: &mut Grid, robot: &mut Coord, d: Coord) {
    let mut seen = FxHashSet::default();

    let mut queue = VecDeque::new();
    queue.push_back(*robot);
    while let Some(pos) = queue.pop_front() {
        if seen.contains(&pos) {
            continue;
        }
        seen.insert(pos);

        let new_pos = pos + d;

        match grid[new_pos] {
            '#' => {
                return;
            }
            '[' => {
                queue.push_back(new_pos);
                queue.push_back(new_pos + Coord::RIGHT);
            }
            ']' => {
                queue.push_back(new_pos + Coord::LEFT);
                queue.push_back(new_pos);
            }
            '.' => (),
            _ => panic!(),
        };
    }

    while !seen.is_empty() {
        let mut seen_new = FxHashSet::default();

        for &pos in &seen {
            let new_pos = pos + d;

            if seen.contains(&new_pos) {
                seen_new.insert(pos);
            } else {
                grid[new_pos] = grid[pos];
                grid[pos] = '.';
            }
        }

        seen = seen_new;
    }

    *robot += d;
}

#[cfg(feature = "anim")]
fn save_warehouse(
    grid: &Grid,
    robot: Coord,
    n: u32,
    moves: usize,
) -> Result<(), Box<dyn std::error::Error>> {
    const SCALE: u32 = 11;

    // limit the frame count
    if moves > 1000 && n % 20 != 0 {
        return Ok(());
    }
    if moves > 100 && n % 5 != 0 {
        return Ok(());
    }

    let width = u32::try_from(grid.width())? * SCALE;
    let height = u32::try_from(grid.height())? * SCALE;

    let mut imgbuf = image::ImageBuffer::new(width, height);

    // Iterate over the coordinates and pixels of the image
    for (x, y, pixel) in imgbuf.enumerate_pixels_mut() {
        let r = u8::try_from(((3 * x) / 10) % 256)?;
        let b = u8::try_from(((3 * y) / 10) % 256)?;
        *pixel = image::Rgb([r, 0, b]);
    }

    {
        let x = u32::try_from(robot.x)? * SCALE;
        let y = u32::try_from(robot.y)? * SCALE;

        for k in 0..((SCALE - 2) * (SCALE - 2)) {
            let pixel = imgbuf.get_pixel_mut(x + 1 + k % (SCALE - 2), y + 1 + k / (SCALE - 2));
            *pixel = image::Rgb([5, 255, 5]);
        }
    }

    for (pos, &c) in grid.iter_cells() {
        let x = u32::try_from(pos.x)? * SCALE;
        let y = u32::try_from(pos.y)? * SCALE;

        if c == '#' {
            for k in 0..(SCALE * SCALE) {
                let pixel = imgbuf.get_pixel_mut(x + k % SCALE, y + k / SCALE);
                *pixel = image::Rgb([240, 10, 70]);
            }
        }

        if c == 'O' {
            for k in 0..((SCALE - 1) * (SCALE - 2)) {
                let pixel = imgbuf.get_pixel_mut(x + 1 + k % (SCALE - 1), y + 1 + k / (SCALE - 2));
                *pixel = image::Rgb([0x9C, 0xDC, 0xFE]);
            }
        }

        if c == '[' {
            for k in 0..((SCALE - 1) * (SCALE - 2)) {
                let pixel = imgbuf.get_pixel_mut(x + 1 + k % (SCALE - 1), y + 1 + k / (SCALE - 2));
                *pixel = image::Rgb([0x9C, 0xDC, 0xFE]);
            }
        }

        if c == ']' {
            for k in 0..((SCALE - 1) * (SCALE - 2)) {
                let pixel = imgbuf.get_pixel_mut(x + k % (SCALE - 1), y + 1 + k / (SCALE - 2));
                *pixel = image::Rgb([0x9C, 0xDC, 0xFE]);
            }
        }
    }

    imgbuf.save(format!("frame{n:05}.png"))?;

    Ok(())
}

struct Puzzle {
    data: String,
    moves: Vec<char>,
}

impl Puzzle {
    fn new(data: &str) -> Self {
        let (a, b) = data.split_once("\n\n").unwrap();

        Self {
            data: a.to_string(),
            moves: b.chars().collect(),
        }
    }

    /// Solve part one.
    fn part1(&self) -> i32 {
        let (mut grid, mut robot) = init_first_warehouse(&self.data);

        #[cfg(feature = "anim")]
        let mut n = 1;

        #[cfg(feature = "anim")]
        let _ = save_warehouse(&grid, robot, 0, 0);

        for m in &self.moves {
            let d = match m {
                '>' => Coord::RIGHT,
                '<' => Coord::LEFT,
                'v' => Coord::DOWN,
                '^' => Coord::UP,
                _ => continue,
            };

            match grid[robot + d] {
                '.' => {
                    robot += d;
                }
                'O' => {
                    let mut i = 1;
                    while grid[robot + i * d] == 'O' {
                        i += 1;
                    }
                    if grid[robot + i * d] == '.' {
                        grid[robot + i * d] = 'O';
                        grid[robot + d] = '.';

                        robot += d;
                    }
                }
                '#' => (),
                _ => panic!(),
            };

            #[cfg(feature = "anim")]
            {
                let _ = save_warehouse(&grid, robot, n, self.moves.len());
                n += 1;
            }
        }

        score(&grid)
    }

    /// Solve part two.
    fn part2(&self) -> i32 {
        let (mut grid, mut robot) = init_second_warehouse(&self.data);

        #[cfg(feature = "anim")]
        let mut n = 1;

        #[cfg(feature = "anim")]
        let _ = save_warehouse(&grid, robot, 0, 0);

        for m in &self.moves {
            let d = match m {
                '>' => Coord::RIGHT,
                '<' => Coord::LEFT,
                'v' => Coord::DOWN,
                '^' => Coord::UP,
                _ => continue,
            };

            match grid[robot + d] {
                '.' => {
                    robot += d;
                }
                'O' | '[' | ']' => {
                    move_boxes(&mut grid, &mut robot, d);
                }
                _ => {}
            }

            #[cfg(feature = "anim")]
            {
                let _ = save_warehouse(&grid, robot, n, self.moves.len());
                n += 1;
            }
        }

        score(&grid)
    }
}

/// # Panics
/// over malformed input
#[must_use]
pub fn solve(data: &str) -> (i32, i32) {
    let puzzle = Puzzle::new(data);
    (puzzle.part1(), puzzle.part2())
}

pub fn main() {
    let args = aoc::parse_args();

    #[cfg(feature = "anim")]
    if args.has_option("--anim1") {
        let puzzle = Puzzle::new(&args.input);
        puzzle.part1();
        std::process::exit(0);
    }

    #[cfg(feature = "anim")]
    if args.has_option("--anim2") {
        let puzzle = Puzzle::new(&args.input);
        puzzle.part2();
        std::process::exit(0);
    }

    args.run(solve);
}

#[cfg(test)]
mod test {
    use super::*;

    const SAMPLE_1: &str = include_str!("sample_1.txt");
    const SAMPLE_2: &str = include_str!("sample_2.txt");

    #[test]
    fn test01() {
        let puzzle = Puzzle::new(SAMPLE_1);
        assert_eq!(puzzle.part1(), 10092);
    }

    #[test]
    fn test02() {
        let puzzle = Puzzle::new(SAMPLE_2);
        assert_eq!(puzzle.part1(), 2028);
    }

    #[test]
    fn test03() {
        let puzzle = Puzzle::new(SAMPLE_1);
        assert_eq!(puzzle.part2(), 9021);
    }
}

File: src/year2024/day15/day15.go
// [Day 15: Warehouse Woes](https://adventofcode.com/2024/day/15)

package main

import (
	"container/list"
	"fmt"
	"log"
	"os"
	"strings"
	"time"
)

// Coord represents a 2D coordinate.
type Coord struct {
	X, Y int
}

var (
	RIGHT = Coord{X: 1, Y: 0}
	LEFT  = Coord{X: -1, Y: 0}
	DOWN  = Coord{X: 0, Y: 1}
	UP    = Coord{X: 0, Y: -1}
)

func (c Coord) add(d Coord) Coord {
	return Coord{X: c.X + d.X, Y: c.Y + d.Y}
}

type Grid struct {
	data     []byte
	width    int
	height   int
	exterior byte
}

func parseGrid(input string, exterior byte) Grid {
	lines := strings.Split(strings.TrimSpace(input), "\n")
	if len(lines) == 0 {
		return Grid{exterior: exterior}
	}

	width := 0
	for _, line := range lines {
		if len(line) > width {
			width = len(line)
		}
	}
	height := len(lines)

	var gridData []byte
	for _, line := range lines {
		gridData = append(gridData, []byte(line)...)
		for len(line) < width {
			gridData = append(gridData, ' ')
			line += " "
		}
	}

	return Grid{
		data:     gridData,
		width:    width,
		height:   height,
		exterior: exterior,
	}
}

func (g Grid) get(p Coord) byte {
	if p.X < 0 || p.X >= g.width || p.Y < 0 || p.Y >= g.height {
		return g.exterior
	}
	return g.data[p.Y*g.width+p.X]
}

func (g *Grid) set(p Coord, c byte) {
	if p.X < 0 || p.X >= g.width || p.Y < 0 || p.Y >= g.height {
		return
	}
	g.data[p.Y*g.width+p.X] = c
}

func (g Grid) withSize(width, height int, value, exterior byte) Grid {
	data := make([]byte, width*height)
	for i := range data {
		data[i] = value
	}
	return Grid{
		data:     data,
		width:    width,
		height:   height,
		exterior: exterior,
	}
}

func score(grid Grid) int {
	sum := 0
	for y := 0; y < grid.height; y++ {
		for x := 0; x < grid.width; x++ {
			c := grid.get(Coord{X: x, Y: y})
			if c == 'O' || c == '[' {
				sum += 100*y + x
			}
		}
	}
	return sum
}

func initFirstWarehouse(input string) (Grid, Coord) {
	grid := parseGrid(input, '#')
	var start Coord

	for y := 0; y < grid.height; y++ {
		for x := 0; x < grid.width; x++ {
			pos := Coord{X: x, Y: y}
			if grid.get(pos) == '@' {
				start = pos
				grid.set(pos, '.')
				break
			}
		}
		if start.X != 0 || start.Y != 0 {
			break
		}
	}

	return grid, start
}

func initSecondWarehouse(input string) (Grid, Coord) {
	simple := parseGrid(input, '#')
	grid := simple.withSize(simple.width*2, simple.height, ' ', '#')
	var start Coord

	for y := 0; y < simple.height; y++ {
		for x := 0; x < simple.width; x++ {
			pos := Coord{X: x, Y: y}
			c := simple.get(pos)
			pos1 := Coord{X: x * 2, Y: y}
			pos2 := Coord{X: x*2 + 1, Y: y}

			switch c {
			case '@':
				start = pos1
				grid.set(pos1, '.')
				grid.set(pos2, '.')
			case 'O':
				grid.set(pos1, '[')
				grid.set(pos2, ']')
			default:
				grid.set(pos1, c)
				grid.set(pos2, c)
			}
		}
	}

	return grid, start
}

func moveBoxes(grid *Grid, robot *Coord, d Coord) {
	seen := make(map[Coord]bool)
	queue := list.New()
	queue.PushBack(*robot)

	for queue.Len() > 0 {
		front := queue.Front()
		queue.Remove(front)
		pos := front.Value.(Coord)

		if seen[pos] {
			continue
		}
		seen[pos] = true

		newPos := pos.add(d)

		switch grid.get(newPos) {
		case '#':
			return
		case '[':
			queue.PushBack(newPos)
			queue.PushBack(newPos.add(RIGHT))
		case ']':
			queue.PushBack(newPos.add(LEFT))
			queue.PushBack(newPos)
		case '.':
			// continue
		}
	}

	for len(seen) > 0 {
		seenNew := make(map[Coord]bool)

		for pos := range seen {
			newPos := pos.add(d)

			if seen[newPos] {
				seenNew[pos] = true
			} else {
				grid.set(newPos, grid.get(pos))
				grid.set(pos, '.')
			}
		}

		seen = seenNew
	}

	*robot = robot.add(d)
}

func part1(data string) int {
	parts := strings.SplitN(data, "\n\n", 2)
	if len(parts) != 2 {
		return 0
	}
	warehouseData := parts[0]
	movesStr := parts[1]

	grid, robot := initFirstWarehouse(warehouseData)

	for _, m := range movesStr {
		var d Coord
		switch m {
		case '>':
			d = RIGHT
		case '<':
			d = LEFT
		case 'v':
			d = DOWN
		case '^':
			d = UP
		default:
			continue
		}

		newPos := robot.add(d)
		switch grid.get(newPos) {
		case '.':
			robot = newPos
		case 'O':
			i := 1
			for grid.get(robot.add(Coord{X: d.X * i, Y: d.Y * i})) == 'O' {
				i++
			}
			finalPos := robot.add(Coord{X: d.X * i, Y: d.Y * i})
			if grid.get(finalPos) == '.' {
				grid.set(finalPos, 'O')
				grid.set(newPos, '.')
				robot = newPos
			}
		case '#':
			// cannot move
		}
	}

	return score(grid)
}

func part2(data string) int {
	parts := strings.SplitN(data, "\n\n", 2)
	if len(parts) != 2 {
		return 0
	}
	warehouseData := parts[0]
	movesStr := parts[1]

	grid, robot := initSecondWarehouse(warehouseData)

	for _, m := range movesStr {
		var d Coord
		switch m {
		case '>':
			d = RIGHT
		case '<':
			d = LEFT
		case 'v':
			d = DOWN
		case '^':
			d = UP
		default:
			continue
		}

		newPos := robot.add(d)
		switch grid.get(newPos) {
		case '.':
			robot = newPos
		case 'O', '[', ']':
			moveBoxes(&grid, &robot, d)
		}
	}

	return score(grid)
}

func solve(data []byte) (int, int) {
	input := string(data)
	return part1(input), part2(input)
}

func main() {
	filename := "input.txt"
	elapsed := false

	for i := 1; i < len(os.Args); i++ {
		arg := os.Args[i]
		if arg == "--elapsed" {
			elapsed = true
		} else if !strings.HasPrefix(arg, "-") {
			filename = arg
		}
	}

	data, err := os.ReadFile(filename)
	if err != nil {
		log.Fatal(err)
	}

	if elapsed {
		start := time.Now()
		result1, result2 := solve(data)
		duration := time.Since(start)

		fmt.Println(result1)
		fmt.Println(result2)
		fmt.Printf("elapsed: %f ms\n", duration.Seconds()*1000.0)
	} else {
		result1, result2 := solve(data)

		fmt.Println(result1)
		fmt.Println(result2)
	}
}
